refactor: standardize Prometheus metric naming conventions (part 1)

keivenchang opened the pull request:
#### Overview:

This PR improves consistency in Prometheus naming conventions and updating related documentation. No functionality has been changed.

#### Details:

• Updated metrics collection and prometheus naming in lib/runtime/src/metrics/prometheus_names.rs
• Updated service metrics handling in lib/runtime/src/service.rs and lib/llm/src/http/service/metrics.rs
• Updated metrics deployment documentation in deploy/metrics/README.md

#### Where should the reviewer start?

Focus on lib/runtime/src/metrics/prometheus_names.rs for the core naming convention changes, then review lib/runtime/src/service.rs for service-level metrics updates.

/coderabbit profile chill

grahamking commented:
Code Rabbit making some interesting points.

keivenchang commented on lib/runtime/examples/system_metrics/README.md:
```diff
@@ -62,7 +62,7 @@ The `dynamo_component_errors_total` metric includes the following error types:
 
 ### Gauges
-- `dynamo_component_inflight_requests` - Number of requests currently being processed
+- `dynamo_component_inflight_requests_total` - Number of requests currently being processed
```
Two human beings think that, for the sake of consistency, having a unit as the suffix (e.g. _total) will be consistent. See other examples in prometheus_names.rs.

keivenchang replied:
I removed _total from gauge. Done.

keivenchang commented on lib/runtime/src/pipeline/network/ingress/push_handler.rs:
```diff
@@ -73,7 +73,7 @@ impl WorkHandlerMetrics {
         )?;
 
         let inflight_requests = endpoint.create_intgauge(
-            work_handler::INFLIGHT_REQUESTS,
+            work_handler::INFLIGHT_REQUESTS_TOTAL,
             "Number of requests currently being processed by work handler",
             metrics_labels,
         )?;
```
How many times does Rabbit need to repeat the same drivel?

michaelfeil commented on lib/llm/src/http/service/metrics.rs:
```diff
@@ -96,7 +96,8 @@ impl Metrics {
     /// The following metrics will be created with the configured prefix:
     /// - `{prefix}_requests_total` - IntCounterVec for the total number of requests processed
-    /// - `{prefix}_inflight_requests` - IntGaugeVec for the number of inflight requests
+    /// - `{prefix}_inflight_requests_total` - IntGaugeVec for the number of inflight requests
+    /// - `{prefix}_client_disconnects_total` - IntGauge for the number of connections dropped by clients
     /// - `{prefix}_request_duration_seconds` - HistogramVec for the duration of requests
```
Fine with me. We have not rolled out the version yet (assuming you talk about the cancelled request counter, answering from mobile)

keivenchang commented on deploy/metrics/README.md:
```diff
@@ -40,7 +40,7 @@ As of Q2 2025, Dynamo HTTP Frontend metrics are exposed when you build container
 The core Dynamo backend system automatically exposes metrics with the `dynamo_component_*` prefix for all components that use the `DistributedRuntime` framework:
 
-- `dynamo_component_inflight_requests`: Requests currently being processed (gauge)
+- `dynamo_component_inflight_requests_total`: Requests currently being processed (gauge)
```
Ok. Removed _total from gauge.

biswapanda reviewed (approved):
lgtm

keivenchang committed:
feat: standardize Prometheus metric names and add comprehensive metrics support

- Update metric names to remove '_total' suffix from gauge metrics
- Add KvStatsMetrics class for KV cache statistics
- Add missing frontend service metrics (disconnected_clients, model config metrics)
- Update Python bindings with new metric name constants
- Fix HTTP metrics test assertions to use updated metric names
- Reorganize metrics documentation and implementation
- Apply consistent formatting across metrics code

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

rmccorm4 commented on lib/llm/tests/http_metrics.rs:
```diff
@@ -386,6 +386,28 @@ mod integration_tests {
+                    if let Err(e) = service
+                        .state()
+                        .metrics_clone()
+                        .update_metrics_from_model_entry_with_mdc(&entry, &etcd_client)
```
Was there still more to the bad merge?

Seeing this failed check:
```
error[E0599]: no method named `update_metrics_from_model_entry_with_mdc` found for struct `Arc<dynamo_llm::http::service::Metrics>` in the current scope
   --> lib/llm/tests/http_metrics.rs:398:26
    |
395 |                       if let Err(e) = service
    |  _____________________________________-
396 | |                         .state()
397 | |                         .metrics_clone()
398 | |                         .update_metrics_from_model_entry_with_mdc(&entry, &etcd_client)
    | |_________________________-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
help: there is a method `update_metrics_from_mdc` with a similar name, but with different arguments
   --> /workspace/lib/llm/src/http/service/metrics.rs:475:5
    |
475 |     pub fn update_metrics_from_mdc(&self, card: &ModelDeploymentCard) -> anyhow::Result<()> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

keivenchang committed:
test: fix http_metrics integration test for MDC metrics validation

Update test_metrics_with_mdc_registration to properly populate model cards and MDC metrics by manually calling save_model_card and update_metrics_from_mdc, simulating the ModelWatcher polling behavior

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>
